fix: (mariadb setup) Issue 5661 - MariaDB 10.3 Barracuda and InnoDB

brian-pond opened the pull request:
This code is my proposed solution for Issue 5661
- MariaDB version 10.3.x did **not** support the config keyword _innodb_file_format_
- Previous versions supported it.
- Version 10.4.x brought the keyword back for compatibility.  **But**...it's not actually used anymore

For users with MariaDB version 10.3.x, the code in _setup_db.py_ threw errors.  Code expected the _innodb_file_format_ keyword. But could not find it.  Because MariaDB doesn't have that keyword in version 10.3.x.

After a ton of trial and error, I have performed a large re-write of  _setup_db.py_.  
- Created a function to calculate the MariaDB major and minor version. MariaDB considers the first 2 numbers to be the Major version. And the 3rd number to be the minor version. This is different than ERPNext, where we have Major.Minor.Patch.
    Source = https://mariadb.com/kb/en/library/upgrading-between-minor-versions-on-linux/
- This function future-fixes bugs.  Someday, MariaDB will move to version 11. Our code _would_ have broken at 11.3, because it only searched for "3". Instead of "10.3". Now it handles all future versions.
- Searched for different global settings, depending on whether MariaDB is 10.2 or earlier.  Or 10.3 and later.
- Combined the 2 validations (barracuda, database) into a single function.  Much easier to read now.
- Removed checks for client-only variables, that you cannot retrieve from MariaDB global variables.
- Cleaned up error messages, so they are clear and easy to understand.
- Applied PEP8 rules to the python file, so that indentation and formatting is very consistent.

I've tested on Ubuntu 18.10, MariaDB 10.4.3, using a fresh installation.

Please contact me if you have any questions!
Brian
brian@pondconsulting.net

brian-pond committed:
Fixed setup_db.py to support MariaDB version 10.3.  Added function for calculating versions, per MariaDB documentation.

brian-pond committed:
Merge remote-tracking branch 'github_frappe/develop' into 5661_Brian

Synchronizing with Frappe's develop branch on GitHub.

brian-pond committed:
Major rewrite of setup_db.py. Reduced number of functions. Made fully compatible with MariaDB 10.3+. Removed references to client parameters in my.cnf that cannot be queried.  Applied PEP8 rules to python code.

Thunderbottom commented on frappe/database/mariadb/setup_db.py:
```diff
@@ -1,148 +1,167 @@
+               "Please verify the settings below in MariaDB's my.cnf.  Restart MariaDB.  And {sep}"
+               "then run `bench new-site {x}` again.{sep2}"
+               "").format(x=site, sep2="\n"*2, sep="\n")
+
```
is there something missing over here?

Thunderbottom replied:
Oh, you're missing `print_db_config()` in here.

PMojito replied:
Ah, you're right!  What's the best way for me to fix that?  Do I need close and re-submit a brand new pull?

Thunderbottom replied:
No, just push more commits to this branch. It should show up here.

PMojito replied:
Excellent, thanks for that info.  :)  I've added the missing line to call print_db_config().

brian-pond committed:
Adding missing print statement in setup_db.py

netchampfaris commented:
@PMojito Hey, thanks for this PR. You will also need to fix tests.
You can run tests locally using the command
```
bench --site sitename run-tests
```

PMojito commented:
No problem!  I will run tests, and get those updated.
Thanks for the help, guys.  I look forward to working with you a lot more!  :smiley:

rmehta commented:
Tabs, please: https://github.com/frappe/erpnext/wiki/Coding-Standards#tabs-not-spaces

Also do read the contribution guidelines!

brian-pond committed:
Replaced 4-spaces with tabs, per ERPNext standards.

PMojito commented:
I've replaced spaces with tabs.  I'd been using _autopep8_ on my Python files.  And pep8 prefers spaces.  It doesn't appear that's easily configurable with autopep8.  So I'll need to research more on the web.  And try to find an alternative.  Probably there's another linter I could substitute.

PMojito commented:
Apologies for the delay on running tests.  It's challenging, because the repository code (on GitHub), is **not** the same as a real, working ERPNext environment.

- GitHub.  My modified code = **..frappe/database/mariadb/setup_db.py**
- A real, working ERPNext directory:  _There is no such file_ **"setup_db.py"**

I _think_ this is what happens.  When bench installs Frappe, it's taking the code within  "setup_db.py" , and  _merging_ into **./apps/frappe/frappe/installer.py**.   I'm not sure why.  But it seems to be true.

So to run tests?  I must manually copy all my changes from setup_db.py.  And write them _again_, inside of installer.py.  Then I could try to run tests.

Thunderbottom commented:
> Probably there's another linter I could substitute.
there's [coala](https://coala.io) if you want.

Thunderbottom commented:
> Apologies for the delay on running tests. It's challenging, because the repository code (on GitHub), is **not** the same as a real, working ERPNext environment.
> 
>     * GitHub.  My modified code = **..frappe/database/mariadb/setup_db.py**
> 
>     * A real, working ERPNext directory:  _There is no such file_ **"setup_db.py"**
> 
> 
> I _think_ this is what happens. When bench installs Frappe, it's taking the code within "setup_db.py" , and _merging_ into **./apps/frappe/frappe/installer.py**. I'm not sure why. But it seems to be true.

If you look at the directory structure, you can find the file to be inside `frappe-bench/apps/frappe/frappe/database/mariadb/setup_db.py`

> So to run tests? I must manually copy all my changes from setup_db.py. And write them _again_, inside of installer.py. Then I could try to run tests.

No need to do this, just run what @netchampfaris suggested in his comment, it should work just fine. If there are any issues with the `mariadb` test (which probably should fail with all these changes), then it will throw at you some errors that'll let you know what's gone wrong while running the test.

Thunderbottom commented:
Also, for your tests, you can check the travis checks on your PR: [here](https://travis-ci.org/frappe/frappe/jobs/506861762) and [here](https://travis-ci.org/frappe/frappe/jobs/506861760). You can see that your test fails here:

```python-traceback
======================================================================
ERROR: test_patch_module_names (frappe.tests.test_patches.TestPatches)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/frappe-bench/apps/frappe/frappe/tests/test_patches.py", line 15, in test_patch_module_names
    self.assertTrue(frappe.get_attr(patchmodule.split()[0] + ".execute"))
  File "/home/travis/frappe-bench/apps/frappe/frappe/__init__.py", line 1018, in get_attr
    return getattr(get_module(modulename), methodname)
  File "/home/travis/frappe-bench/apps/frappe/frappe/__init__.py", line 799, in get_module
    return importlib.import_module(modulename)
  File "/opt/python/2.7.14/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/travis/frappe-bench/apps/frappe/frappe/patches/v5_0/convert_to_barracuda_and_utf8mb4.py", line 3, in <module>
    from frappe.database.mariadb.setup_db import check_if_ready_for_barracuda
ImportError: cannot import name check_if_ready_for_barracuda
```

PMojito commented:
Chinmay, there is only a "__pycache__" file in that directory.  Here's a screenshot.

![image](https://user-images.githubusercontent.com/19827963/54465160-f83c0580-4736-11e9-9ea5-892d4983c623.png)

I don't know why this looks different, versus just cloning what's on GitHub.

Commit:
Fixed broken test by substituing 'check_database_settings()'

Fixed the test in convert_to_barracuda_and_utf8mb4.py by calling the new function 'check_database_settings()'

PMojito commented:
Also, my "installer.py" is larger than the one on GitHub.  And it has more functions.

Bench is modifying this during installation.  I just don't understand _why_ it's doing this.  Any bench experts on Frappe team who can explain?

Thunderbottom commented:
Hm, I am really unsure as to why that is happening for you. It looks the same across all the benches I have.

Commit:
Merge pull request #4 from frappe/develop

Sync from Frappe - 4 April 2019

Thunderbottom commented:
Has this been tested @PMojito ?

PMojito commented:
Hi Chinmay - I will re-test today, using latest from development branch.  Standby...

Thunderbottom commented:
Sure, I'll have it tested locally as well.

brian-pond committed:
Merge branch 'develop' into 5661_Brian

brian-pond committed:
Merge branch 'develop' of github.com:frappe/frappe into 5661_Brian

brian-pond committed:
Cleaned up setup.py, to match latest version of Frappe develop

PMojito commented:
Chinmay, I've synchronized 1 more time, and reviewed my code changes.  I think this is ready.
